onnxruntime/core/session/ep_plugin_provider_interfaces.cc

```diff
@@ -0,0 +1,270 @@
+                                               const OrtLogger& session_logger) {
+  OrtEp* ort_ep = nullptr;
+  UniqueOrtStatus status(ep_factory_.CreateEp(&ep_factory_, devices_.data(), ep_metadata_.data(), devices_.size(),
+                                              &session_options, &session_logger, &ort_ep),
+                         OrtApis::ReleaseStatus);
+
+  if (status != nullptr) {
+    ORT_THROW("Error creating execution provider: ", ToStatus(status.get()).ToString());
+  }
+
+  auto ep_wrapper = std::make_unique<PluginExecutionProvider>(UniqueOrtEp(ort_ep, OrtEpDeleter(ep_factory_)));
+  ep_wrapper->SetLogger(session_logger.ToInternal());
+
+  return ep_wrapper;
+}
+
+/// <summary>
+/// Functor used to generate a Metadef name for a subgraph supported by a plugin EP.
+/// The generated name is a concatenation of a prefix (i.e., the EP name) with
+/// the model's hash and a unique ID.
+/// </summary>
+struct PluginEpMetaDefNameFunctor {
+  explicit PluginEpMetaDefNameFunctor(const ModelMetadefIdGenerator& generator,
+                                      const GraphViewer& graph_viewer,
+                                      const std::string& prefix)
+      : generator_(generator), graph_viewer_(graph_viewer), prefix_(prefix) {}
+
+  std::string operator()() {
+    uint64_t model_hash = 0;
+    int id = generator_.GenerateId(graph_viewer_, model_hash);
+    return MakeString(prefix_, "_", model_hash, "_", id);
+  }
+
+  const ModelMetadefIdGenerator& generator_;
+  const GraphViewer& graph_viewer_;
+  const std::string& prefix_;
+};
+
+//
+// PluginExecutionProvider
+//
+
+PluginExecutionProvider::PluginExecutionProvider(UniqueOrtEp ep)
+    : IExecutionProvider(ep->GetName(ep.get()), OrtDevice()),  // TODO: What to do about OrtDevice for plugins?
+      ort_ep_(std::move(ep)) {
+}
+
+PluginExecutionProvider::~PluginExecutionProvider() {
+  if (ort_ep_ && !api_node_compute_infos_.empty()) {
+    ort_ep_->ReleaseNodeComputeInfos(ort_ep_.get(), api_node_compute_infos_.data(),
+                                     api_node_compute_infos_.size());
+  }
+}
+
+std::vector<std::unique_ptr<ComputeCapability>>
+PluginExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph_viewer,
+                                       const IKernelLookup& kernel_lookup,
+                                       const GraphOptimizerRegistry& graph_optimizer_registry,
+                                       IResourceAccountant* resource_accountant) const {
+  ORT_UNUSED_PARAMETER(graph_optimizer_registry);  // TODO: Add support
+  ORT_UNUSED_PARAMETER(resource_accountant);       // TODO: Add support? Not used by prioritized EPs
+  ORT_UNUSED_PARAMETER(kernel_lookup);             // TODO: Add support? Not used by prioritized EPs, so probably not needed?
+
+  std::unique_ptr<EpGraph> ep_graph = nullptr;
+  if (Status status = EpGraph::Create(graph_viewer, /*parent_node*/ nullptr, ep_graph); !status.IsOK()) {
```

Comment:
We don't handle parent node for now?

Reply:
I'm not properly accounting for GetCapability's bottom-up subgraph traversal. I'll fix it. Thanks for pointing this out.

Reply:
We don't need to pass this in. I removed that parameter.